### ISSUE TITLE
[22] Byte code backward compatibility for compilation targets < 22 violated with '-proceedOnError', because parser falsely identifies unnamed class

### ISSUE THREAD

**Issue by kriegaex** [eclipse-jdt/eclipse.jdt.core#2205]
Consider this code with a syntax error:

```java
blahblahpackage pack;

public class C {
  public static void main(String[] argv) {}
}
```

```java
public class Invoker {
  public static void main(String[] args) throws Throwable {
    try {
      C.main(null);
    }
    catch (Throwable t) {
      boolean failedCorrectly = t.toString().indexOf("Unresolved compilation") != -1;
      if (failedCorrectly)
        return;
      throw new RuntimeException("Call to main should have failed!", t);
    }
  }
}
```

Now compile with ECJ 3.36.0 and `-8 -proceedOnError` (I also tried with `-1.5`) to cause the compiler to render error messages into the byte code:

```text
$ java -jar ecj-3.36.0.jar -8 -proceedOnError Invoker.java C.java
----------                                                                                        
1. ERROR in C:\Users\alexa\Documents\java-src\AspectJ\tests\bugs161\pr102733\_x\C.java (at line 1)
        blahblahpackage pack;                                                                     
        ^^^^^^^^^^^^^^^                                                                           
Syntax error on token "blahblahpackage", package expected                                         
----------                                                                                        
1 problem (1 error)                                                                               

alexa@Xander-UB MINGW64 ~/Documents/java-src/AspectJ/tests/bugs161/pr102733/_x (java-22)
$ java -cp . Invoker
```

You see, `Invoker` prints nothing, because the generated byte code is as expected.

Now, try the same with an ECJ or batch compiler snapshot after branch `BETA_JAVA22` was merged onto the main branch (I tried with commit b2705b00e91be7b06d36d75bdd6e052450991021):

```text
$ java -jar org.eclipse.jdt.core.compiler.batch-3.38.0-SNAPSHOT.jar -8 -proceedOnError Invoker.java C.java
----------
1. ERROR in C:\Users\alexa\Documents\java-src\AspectJ\tests\bugs161\pr102733\_x\Invoker.java (at line 4)
        C.main(null);
          ^^^^
The method main(null) is undefined for the type C
----------
2. ERROR in C:\Users\alexa\Documents\java-src\AspectJ\tests\bugs161\pr102733\_x\C.java (at line 1)
        blahblahpackage pack;
        ^
The preview feature Unnamed Classes and Instance Main Methods is only available with source level 22 and above
----------
3. ERROR in C:\Users\alexa\Documents\java-src\AspectJ\tests\bugs161\pr102733\_x\C.java (at line 1)
        blahblahpackage pack;
        ^^^^^^^^^^^^^^^
blahblahpackage cannot be resolved to a type
----------
4. ERROR in C:\Users\alexa\Documents\java-src\AspectJ\tests\bugs161\pr102733\_x\C.java (at line 1)
        blahblahpackage pack;
        ^
Implicitly declared class must have a candidate main method
----------
5. ERROR in C:\Users\alexa\Documents\java-src\AspectJ\tests\bugs161\pr102733\_x\C.java (at line 4)
        public static void main(String[] argv) {}
                           ^^^^^^^^^^^^^^^^^^^
The method main cannot be declared static; static methods can only be declared in a static or top level type
----------
5 problems (5 errors)

$ java -cp . Invoker
Exception in thread "main" java.lang.Error: Unresolved compilation problem: 
        The method main(null) is undefined for the type C

        at Invoker.main(Invoker.java:4)
```

You see that despite compiling with a target less than 22, there is the ***"The preview feature Unnamed Classes and Instance Main Methods is only available with source level 22 and above"*** error. The byte code also looks rather like what would be expected in Java 22, because it creates a class `C` instead of what it thinks is an unnamed class and also an inner class `C.C` (albeit, not static).

In a nutshell: The byte code generated by 3.38-SNAPSHOT breaks compatibility, because both structure and behaviour are different than in 3.36.0.

I cannot compare the behaviour to Javac, because it does not have a feature similar to `-proceedOnError`.

**Comment by kriegaex** [comment]
Just in case you are wondering how I found such a weird example: In AspectJ, an old regression test ensuring compatibility between AJC and ECJ started to fail, because ECJ now behaves differently than before.

**Comment by stephan-herrmann** [comment]
This is a tricky one. 

First observation: our grammar lacks suitable `/:$compliance 22:/` declarations for this feature, but even with that, the example is parsed as an implicitly declared class. Only DiagnoseParser uses that information in order to produce suitable error messages (see usage of `Parser.rules_compliance`).

The core problem is: syntactically `blahblahpackage pack;` leaves no doubt that this is a variable declaration, and since we are not in a method context it must be a field declaration, which can only happen in a type declaration, explicit or implicit.

some ideas:
(1) should `Parser.parse()` directly check `act` similar to how DiagnoseParser does it? But then what?

(2) should we insert an empty token to the grammar, like:
```
ImplicitlyDeclaredClassBodyDeclarations ::= ClassMemberDeclaration StartImplicitType ImplicitlyDeclaredClassBodyDeclarations
```

Then we could let `StartImplicitType` trigger a consume method that does - what?
* discard the previously parsed ClassMemberDeclaration?
* reset stacks? which? how to ensure a consistent state?

(3) perhaps it's enough to make `Parser.consumeInternalCompilationUnitWithPotentialImplicitlyDeclaredClass()` conditional like this:
* if the feature is not enabled, then
  * if we find exactly one type declaration, we could promote this to the toplevel of the CU, and then
     * report all fields and methods found outside the class as misplaced constructs

@mickaelistria  as the author of `Parser.consumeInternalCompilationUnitWithPotentialImplicitlyDeclaredClass()`, do you want to follow up along the lines of idea (3)?

**Comment by mickaelistria** [comment]
Sorry, I don't plan to work on it on short term.
